### DISCUSSION TITLE
`config` source of truth?

### DISCUSSION THREAD

**Discussion by llllvvuu** [wevm/wagmi#2666] [General]
Some things I observed when I stumbled upon #2665:

1. The `config` which is passed into `WagmiConfig` is not used much. Most of the hooks just call `@wagmi/core`.
2. The reason we can just call `@wagmi/core` outside of the React tree, and even call with no `config` variable, is because `createConfig` (both the React version and the core version) has the side effect of setting a global variable. This global variable is the source of truth. This side-effect is non-obvious since given the name `createConfig` and it's not documented.

While it's nice that we can do 2., it comes at the cost of the pitfalls of global variables (#2665) and not being the React way. Why not either go all the way and have no Provider, or go all the way and use `useConfig()` everywhere as the source of truth (the idiomatic React way / developer expectation)?

One way to incrementally adopt `useConfig()` could be to add an optional `wagmiConfig` parameter to the `@wagmi/core` functions so that lines like `const config = getConfig()` become `const config = wagmiConfig ?? getConfig()`. Then the hooks can call like:

```typescript
const config = useConfig()
const watchAccount = useCallback(callback => watchAccountCore(callback, undefined, config), [config])
```
^this would be mostly backwards-compatible, though technically still breaking, because JavaScript lets you pass in extra junk parameters and some users could be doing that

One could then go further and make a more breaking change like:

* rename `@wagmi/core`'s `createConfig` to `setConfig` and document the side effect
* keep `wagmi` (React)'s `createConfig` as `createConfig` but remove the side effect (React users can then no longer call vanilla `@wagmi/core` without passing in `config`)

Thoughts? Open to the idea? Open to contributions?

**Comment by tmm** [comment]
Nice spot! This is something we are working on addressing in v2.
